### PR TITLE
Add iterator chain from Iterators.jl

### PR DESCRIPTION
**Pull request by ghost**
Adding iterators was previously discussed in #10162.  However, 'chain' was not included and it is among the features that I miss the most coming from Python and its `itertools`.  All of the code was lifted over from [Iterators.jl](https://github.com/JuliaLang/Iterators.jl), I only made some minor modifications to the documentation.  As always, feedback is most welcome.

### CODE REVIEW

File: test/functional.jl [hunk 1]
```diff
@@ -138,3 +138,8 @@ let i = 0
         i <= 10 || break
     end
 end
+
+# chain
+# -----
+
+@test collect(chain(1:2:5, 0.2:0.1:1.6)) == [1:2:5, 0.2:0.1:1.6]
```
Comment (hayd) [inline test/functional.jl]: This should be `[1:2:5; 0.2:0.1:1.6]` (to avoid deprecation message.)

Comment (ninjin) [inline test/functional.jl]: Excellent point, patching and pushing.

**Comment by JeffBezanson** [comment]
The reason I didn't add this one is that it's too slow. I only wanted the iterators that are basically trivial wrappers, so that there's no overhead over writing the code some other way.

**Comment by ninjin** [comment]
Maybe there is an issue in how I approach problems, a part of me where what is "Pythonic" needs to be stomped out.  What I generally use `chain

**Comment by stevengj** [comment]
If something like this is included, it seems like it should accept an iterator of iterators, not just a `Vector`, as discussed in JuliaLang/Iterators.jl#50; maybe it should go into the Iterators package, though.

**Comment by Keno** [comment]
The Flatten iterator with this functionality is in Base now.

**Pull request closed without merging**